api/v1alpha1/linodevpc_webhook.go

```diff
@@ -0,0 +1,266 @@
+	linodevpclog.Info("validate update", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object update.
+	return nil, nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *LinodeVPC) ValidateDelete() (admission.Warnings, error) {
+	linodevpclog.Info("validate delete", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object deletion.
+	return nil, nil
+}
+
+func (r *LinodeVPC) validateLinodeVPC(ctx context.Context, client LinodeClient) error {
+	var errs field.ErrorList
+
+	if err := r.validateLinodeVPCSpec(ctx, client); err != nil {
+		errs = slices.Concat(errs, err)
+	}
+
+	if len(errs) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "infrastructure.cluster.x-k8s.io", Kind: "LinodeVPC"},
+		r.Name, errs)
+}
+
+func (r *LinodeVPC) validateLinodeVPCSpec(ctx context.Context, client LinodeClient) field.ErrorList {
+	var errs field.ErrorList
+
+	if err := validateRegion(ctx, client, r.Spec.Region, field.NewPath("spec").Child("region"), LinodeVPCCapability); err != nil {
+		errs = append(errs, err)
+	}
+	if err := r.validateLinodeVPCSubnets(); err != nil {
+		errs = slices.Concat(errs, err)
+	}
+
+	if len(errs) == 0 {
+		return nil
+	}
+	return errs
+}
+
+func (r *LinodeVPC) validateLinodeVPCSubnets() field.ErrorList {
+	var (
+		errs    field.ErrorList
+		builder netipx.IPSetBuilder
+		cidrs   = &netipx.IPSet{}
+		labels  = []string{}
+	)
+
+	for i, subnet := range r.Spec.Subnets {
+		var (
+			label     = subnet.Label
+			labelPath = field.NewPath("spec").Child("Subnets").Index(i).Child("Label")
+			ip        = subnet.IPv4
+			ipPath    = field.NewPath("spec").Child("Subnets").Index(i).Child("IPv4")
+		)
+
+		// Validate Subnet Label
+		if err := validateVPCLabel(label, labelPath); err != nil {
+			errs = append(errs, err)
+		} else if slices.Contains(labels, label) {
+			errs = append(errs, field.Invalid(labelPath, label, "must be unique among the vpc's subnets"))
+		} else {
+			labels = append(labels, label)
+		}
+
+		// Validate Subnet IP Address Range
+		cidr, ferr := validateSubnetIPv4CIDR(ip, ipPath)
+		if ferr != nil {
+			errs = append(errs, ferr)
+			continue
+		}
+		if cidrs.Overlaps(cidr) {
+			errs = append(errs, field.Invalid(ipPath, ip, "range must not overlap with other subnets on the same vpc"))
+		}
+		var err error
+		builder.AddSet(cidr)
+		if cidrs, err = builder.IPSet(); err != nil {
+			return append(field.ErrorList{}, field.InternalError(ipPath, fmt.Errorf("build ip set: %w", err)))
+		}
+	}
+
+	if len(errs) == 0 {
+		return nil
+	}
+	return errs
+}
+
+// TODO: Replace the OpenAPI schema validation for .metadata.name.
+//
+// validateVPCLabel validates a label string is a valid [Linode VPC Label].
+//
+// [Linode VPC Label]: https://www.linode.com/docs/api/vpcs/#vpc-create__request-body-schema
+func validateVPCLabel(label string, path *field.Path) *field.Error {
```

Comment:
could this be done in a similar way to the vpc name? https://github.com/linode/cluster-api-provider-linode/blob/main/config/crd/patches/validation_in_linodevpcs.yaml

Reply:
yes, I understand they are different, my question is could we do this validation as part of the CRD validation rule as well or does this need to be a check in the webhook specifically?

Reply:
Perhaps, but I think the reverse (consolidating the validations with the definitions) would be the better approach here.

Reply:
can you elaborate a bit on why that would be the better approach? it feels to me like minimizing the amount of code we have to maintain and test by using simple regex and other CRD checks would be a benefit

Reply:
1. The validations would be "next to" the definitions in the code. This way if our custom resource definition changes or any specific validations on those definitions change, we can match each API version with the appropriate corresponding validations.

2. The Kubernetes Validation Rules feature is a separate mechanism from the validating admission webhooks, which resources would need to pass two validation gates with different rules, messages, etc. One of the goals of this feature(s) was to surface user configuration errors as early and thoroughly as possible. I believe the split implementation goes against this goal.

Reply:
both implementations would return an error immediately to the user right? or do the validation rules return errors in a different way?

Reply:
The issue here is that it's a _separate_ mechanism and thus they're completely _separate_ validation checks. For example, if the user violated an `x-kubernetes-validations` rule, the API server would return the error early and not call the corresponding validating admission webhook at all.

Reply:
ah, so if we put it all in webhooks users would get all of the errors at once rather than having to fix some of them and then moving on to getting the webhook errors?

Reply:
Yes, I believe that's a better user experience overall (within limitations). For example, we still use things like [controller-gen markers](https://book.kubebuilder.io/reference/markers/crd-validation#crd-validation) to do the bulk of our validations via `x-kubernetes-validations`.